### ISSUE TITLE
[Phase 2.6] Optimizer の実装

### ISSUE THREAD

**Issue by Showhey798** [Showhey798/gradflow#12]
## 概要

パラメータ最適化のための Optimizer を実装します。

## 実装項目

- [ ] `Optimizer` 基底クラス
- [ ] SGD: モーメンタム付き
- [ ] Adam: adaptive learning rate
- [ ] AdamW: weight decay 付き Adam

## ファイル

- `include/gradflow/optim/optimizer.hpp`
- `include/gradflow/optim/sgd.hpp`
- `include/gradflow/optim/adam.hpp`
- `tests/test_optimizer.cpp`

## テスト項目

- [ ] OptimizerTest::SGDConvergence
- [ ] OptimizerTest::AdamConvergence

## 完了基準

- [ ] 簡単な最適化問題（二次関数など）で収束することを確認
- [ ] SGD と Adam の両方が動作

## 関連ドキュメント

- [ROADMAP.md Phase 2.6](docs/ROADMAP.md)

## 期間

Week 5-6

## 依存タスク

- Depends on #11
